crates/red_knot_python_semantic/src/types/call/bind.rs

```diff
@@ -0,0 +1,365 @@
+                    continue;
+                };
+                (index, parameter, ty)
+            }
+
+            Argument::Variadic(_) | Argument::Keywords(_) => {
+                // TODO
+                continue;
+            }
+        };
+        let expected_ty = parameter.annotated_ty();
+        if !argument_ty.is_assignable_to(db, expected_ty) {
+            errors.push(CallBindingError::InvalidArgumentType {
+                parameter: ParameterContext::new(parameter, index),
+                argument_index,
+                expected_ty,
+                provided_ty: *argument_ty,
+            });
+        }
+        if let Some(existing) = parameter_tys[index].replace(*argument_ty) {
+            if parameter.is_variadic() {
+                let union = UnionType::from_elements(db, [existing, *argument_ty]);
+                parameter_tys[index].replace(union);
+            } else {
+                errors.push(CallBindingError::ParameterAlreadyAssigned {
+                    argument_index,
+                    parameter: ParameterContext::new(parameter, index),
+                });
+            }
+        }
+    }
+    if let Some(first_excess_argument_index) = first_excess_positional {
+        errors.push(CallBindingError::TooManyPositionalArguments {
+            first_excess_argument_index,
+            expected_positional_count: signature.positional_parameter_count(),
+            provided_positional_count: next_positional,
+        });
+    }
+    let mut missing = vec![];
+    for (index, bound_ty) in parameter_tys.iter().enumerate() {
+        if bound_ty.is_none() {
+            let param = signature
+                .parameter_at_index(index)
+                .expect("parameter_tys array should not be larger than number of parameters");
+            if param.is_variadic() || param.is_keywords() || param.default_ty().is_some() {
+                // variadic/keywords and defaulted arguments are not required
+                continue;
+            }
+            missing.push(ParameterContext::new(param, index));
+        }
+    }
+
+    if !missing.is_empty() {
+        errors.push(CallBindingError::MissingArguments {
+            parameters: ParameterContexts(missing),
+        });
+    }
+
+    CallBinding {
+        callable_ty,
+        return_ty: signature.return_ty,
+        parameter_tys: parameter_tys
+            .into_iter()
+            .map(|opt_ty| opt_ty.unwrap_or(Type::Unknown))
+            .collect(),
+        errors,
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct CallBinding<'db> {
+    /// Type of the callable object (function, class...)
+    callable_ty: Option<Type<'db>>,
+
+    /// Return type of the call.
+    return_ty: Type<'db>,
+
+    /// Bound types for parameters, in parameter source order.
+    parameter_tys: Box<[Type<'db>]>,
+
+    /// Call binding errors, if any.
+    // TODO use SmallVec once variance bug is fixed
+    errors: Vec<CallBindingError<'db>>,
```

Comment:
I think using a `Vec` here is fine considering that errors shouldn't be that common. We could even consider boxing `errors` to reduce the stack size of `CallBinding`

Reply:
Ok, I can remove the TODO comment.

Not going to add the extra layer of indirection in this PR, I think that should be motivated separately by profiling showing it's a useful tradeoff.

(If overall memory usage of `CallBinding` is an issue, we may want to consider limiting the assigned parameter types that we store, or only storing parameter types at all when the callable is a known-function. We only actually need the assigned parameter types in the binding when we are special-casing a known callable.)